crates/libcontainer/src/process/memory_policy.rs

```diff
@@ -0,0 +1,542 @@
+        // Memory policy mode constants from Linux UAPI (include/uapi/linux/mempolicy.h)
+        const MPOL_DEFAULT: i32 = 0;
+        const MPOL_PREFERRED: i32 = 1;
+        const MPOL_BIND: i32 = 2;
+        const MPOL_INTERLEAVE: i32 = 3;
+        const MPOL_LOCAL: i32 = 4;
+        const MPOL_PREFERRED_MANY: i32 = 5;
+        const MPOL_WEIGHTED_INTERLEAVE: i32 = 6;
+
+        // Memory policy flag constants from Linux UAPI
+        const MPOL_F_NUMA_BALANCING: u32 = 1 << 13; // 0x2000
+        const MPOL_F_RELATIVE_NODES: u32 = 1 << 14; // 0x4000
+        const MPOL_F_STATIC_NODES: u32 = 1 << 15; // 0x8000
+
+        let base_mode = match policy.mode() {
+            MemoryPolicyModeType::MpolDefault => MPOL_DEFAULT,
+            MemoryPolicyModeType::MpolPreferred => MPOL_PREFERRED,
+            MemoryPolicyModeType::MpolBind => MPOL_BIND,
+            MemoryPolicyModeType::MpolInterleave => MPOL_INTERLEAVE,
+            MemoryPolicyModeType::MpolLocal => MPOL_LOCAL,
+            MemoryPolicyModeType::MpolPreferredMany => MPOL_PREFERRED_MANY,
+            MemoryPolicyModeType::MpolWeightedInterleave => MPOL_WEIGHTED_INTERLEAVE,
+        };
+
+        let mut flags_value: u32 = 0;
+        if let Some(flags) = policy.flags() {
+            let mut has_static = false;
+            let mut has_relative = false;
+
+            for flag in flags {
+                match flag {
+                    oci_spec::runtime::MemoryPolicyFlagType::MpolFNumaBalancing => {
+                        // MPOL_F_NUMA_BALANCING is only valid with MPOL_BIND
+                        if base_mode != MPOL_BIND {
+                            return Err(MemoryPolicyError::IncompatibleFlagMode(
+                                "MPOL_F_NUMA_BALANCING can only be used with MPOL_BIND".to_string(),
+                            ));
+                        }
+                        flags_value |= MPOL_F_NUMA_BALANCING;
+                    }
+                    oci_spec::runtime::MemoryPolicyFlagType::MpolFRelativeNodes => {
+                        has_relative = true;
+                        flags_value |= MPOL_F_RELATIVE_NODES;
+                    }
+                    oci_spec::runtime::MemoryPolicyFlagType::MpolFStaticNodes => {
+                        has_static = true;
+                        flags_value |= MPOL_F_STATIC_NODES;
+                    }
+                }
+            }
+
+            // STATIC_NODES and RELATIVE_NODES are mutually exclusive
+            if has_static && has_relative {
+                return Err(MemoryPolicyError::MutuallyExclusiveFlags(
+                    "MPOL_F_STATIC_NODES and MPOL_F_RELATIVE_NODES are mutually exclusive"
+                        .to_string(),
+                ));
+            }
+        }
+
+        // Combine mode and flags
+        let mode_with_flags = base_mode | (flags_value as i32);
+
+        // Handle special cases for MPOL_DEFAULT and MPOL_LOCAL
+        match base_mode {
```

Comment:
Personally, I think it would be clearer to separate validation from the application of set_mempolicy.